File: windows/MoDi.App.Contracts/Connectivity/ChannelContracts.cs
namespace MoDi.App.Contracts.Connectivity;

public enum ChannelRuntimeState { Closed, Open }

public sealed record ChannelDescriptor(ChannelId Id, ChannelKind Kind, ChannelDirection Direction);

public static class AudioChannel
{
    public static readonly ChannelId Primary = ChannelId.Parse("audio/primary");
    public static readonly ChannelDescriptor Descriptor = new(Primary, ChannelKind.Audio, ChannelDirection.Duplex);
}

public interface IChannelDataPlane
{
    SessionId SessionId { get; }
    ChannelDescriptor Descriptor { get; }
    ChannelRuntimeState State { get; }
    long NextSequence { get; }
    event Action<ReadOnlyMemory<byte>>? BytesReceived;
    Task OpenAsync(CancellationToken cancellationToken);
    long ReserveSequence();
    void ResetSequence();
    Task SendAsync(ReadOnlyMemory<byte> payload, CancellationToken cancellationToken);
    Task CloseAsync(CancellationToken cancellationToken);
}

public sealed class ChannelRouter
{
    private readonly object _gate = new();
    private readonly Dictionary<(SessionId Session, ChannelId Channel), IChannelDataPlane> _channels = [];

    public IReadOnlyList<IChannelDataPlane> Channels
    {
        get { lock (_gate) return _channels.Values.ToArray(); }
    }

    public void Register(IChannelDataPlane channel)
    {
        ArgumentNullException.ThrowIfNull(channel);
        lock (_gate)
        {
            if (!_channels.TryAdd((channel.SessionId, channel.Descriptor.Id), channel))
                throw new InvalidOperationException("Channel is already registered for this session.");
        }
    }

    public bool TryGet(SessionId sessionId, ChannelId channelId, out IChannelDataPlane? channel)
    {
        lock (_gate) return _channels.TryGetValue((sessionId, channelId), out channel);
    }

    public async Task CloseSessionAsync(SessionId sessionId, CancellationToken cancellationToken)
    {
        IChannelDataPlane[] matches;
        lock (_gate)
        {
            matches = _channels.Where(pair => pair.Key.Session == sessionId).Select(pair => pair.Value).ToArray();
            foreach (var channel in matches) _channels.Remove((channel.SessionId, channel.Descriptor.Id));
        }
        foreach (var channel in matches) await channel.CloseAsync(cancellationToken).ConfigureAwait(false);
    }
}

File: android/app/src/main/java/com/modi/connect/core/connectivity/ChannelContracts.kt
package com.modi.connect.core.connectivity

enum class ChannelRuntimeState { Closed, Open }

data class ChannelDescriptor(val id: ChannelId, val kind: ChannelKind, val direction: ChannelDirection)

object AudioChannel {
    val primary = ChannelId.parse("audio/primary")
    val descriptor = ChannelDescriptor(primary, ChannelKind.Audio, ChannelDirection.Duplex)
}

object ClipboardChannel {
    val primary = ChannelId.parse("clipboard/main")
    val descriptor = ChannelDescriptor(primary, ChannelKind.Clipboard, ChannelDirection.Duplex)
}

interface ChannelDataPlane {
    val sessionId: SessionId
    val descriptor: ChannelDescriptor
    val state: ChannelRuntimeState
    val nextSequence: Long
    var onBytesReceived: ((ByteArray) -> Unit)?
    suspend fun open()
    fun reserveSequence(): Long
    fun resetSequence()
    suspend fun send(payload: ByteArray)
    suspend fun close()
}

class ChannelRouter {
    private val lock = Any()
    private val entries = linkedMapOf<Pair<SessionId, ChannelId>, ChannelDataPlane>()
    val channels: List<ChannelDataPlane> get() = synchronized(lock) { entries.values.toList() }

    fun register(channel: ChannelDataPlane) = synchronized(lock) {
        check(entries.putIfAbsent(channel.sessionId to channel.descriptor.id, channel) == null) {
            "Channel is already registered for this session"
        }
    }

    fun find(sessionId: SessionId, channelId: ChannelId): ChannelDataPlane? =
        synchronized(lock) { entries[sessionId to channelId] }

    suspend fun closeSession(sessionId: SessionId) {
        val matches = synchronized(lock) {
            entries.filterKeys { it.first == sessionId }.values.toList().also { channels ->
                channels.forEach { entries.remove(it.sessionId to it.descriptor.id) }
            }
        }
        matches.forEach { it.close() }
    }
}
